src/StructureMap.Microsoft.DependencyInjection/AspNetConstructorSelector.cs

```diff
@@ -9,13 +9,26 @@ namespace StructureMap
     internal class AspNetConstructorSelector : IConstructorSelector
     {
         // ASP.NET expects registered services to be considered when selecting a ctor, SM doesn't by default.
-        public ConstructorInfo Find(Type pluggedType, DependencyCollection dependencies, PluginGraph graph) =>
-            pluggedType.GetTypeInfo()
-                .DeclaredConstructors
-                .Select(ctor => new { Constructor = ctor, Parameters = ctor.GetParameters() })
-                .Where(x => x.Parameters.All(param => graph.HasFamily(param.ParameterType)))
-                .OrderByDescending(x => x.Parameters.Length)
-                .Select(x => x.Constructor)
-                .FirstOrDefault();
+        public ConstructorInfo Find(Type pluggedType, DependencyCollection dependencies, PluginGraph graph)
+        {
+            var constructors = from constructor in pluggedType.GetConstructors()
+                               select new
```

Comment:
The fix is the change of:

    Where(x => x.Parameters.All(param => graph.HasFamily(param.ParameterType)))

to 

    where constructor.Parameters.All(parameter => ParameterIsRegistered(parameter, dependencies, graph))

    private static bool ParameterIsRegistered(ParameterInfo parameter, DependencyCollection dependencies, PluginGraph graph)
    {
        return graph.HasFamily(parameter.ParameterType) || dependencies.Any(dependency => dependency.Type == parameter.ParameterType);
    }

We check if we have any dependency in the `dependencies` collection that matches the current parameter, not just if we have the service registered in the container. This way inline dependencies are taken into account when selecting the longest satisfiable constructor.

In the example above, the current code would select the default, parameterless constructor for `MongoClient`. The proposed change would correctly select the constructor that takes a single argument of type `MongoUrl`. For reference, this is the list of constructors for `MongoClient`:
http://api.mongodb.com/csharp/2.2/html/Overload_MongoDB_Driver_MongoClient__ctor.htm